olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -545,95 +487,67 @@ HttpResponse OlpClient::OlpClientImpl::CallApi(
+                                       header_params, post_body, content_type);
 
-  custom_user_agent += http::kOlpSdkUserAgent;
-  network_request.WithHeader(http::kUserAgentHeader, custom_user_agent);
+  PendingUrlRequestPtr request_ptr = nullptr;
+  auto& pending_requests = pending_requests_;
+  const auto& url = network_request->GetUrl();
 
-  if (!content_type.empty()) {
-    network_request.WithHeader(http::kContentTypeHeader, content_type);
-  }
+  struct ResponseData {
+    Condition condition;
+    HttpResponse response;
+  };
+  auto response_data = std::make_shared<ResponseData>();
 
-  auto backdown_period =
-      std::chrono::milliseconds(retry_settings.initial_backdown_period);
+  olp::client::NetworkAsyncCallback callback = [&](HttpResponse http_response) {
+    response_data->response = std::move(http_response);
+    response_data->condition.Notify();
+  };
 
-  AddBearer(query_params.empty(), network_request);
+  // Only merge same request in case there is no body as a body can alter the
+  // outcome of the request and may not match the response of a request with a
+  // different body.
+  bool merge = (!post_body || post_body->empty());
+  OLP_SDK_LOG_DEBUG_F(kLogTag, "CallApi: url='%s', merge='%s'", url.c_str(),
+                      merge ? "true" : "false");
 
-  auto response =
-      SendRequest(network_request, settings_, retry_settings, context);
+  request_ptr = std::make_shared<PendingUrlRequest>(context);
 
-  NetworkStatistics accumulated_statistics = response.GetNetworkStatistics();
+  auto call_id = merge ? pending_requests->Append(url, request_ptr, callback)
+                       : request_ptr->Append(std::move(callback));
 
-  // Make sure that we don't wait longer than `timeout` in retry settings
-  auto accumulated_wait_time = backdown_period;
-  const auto max_wait_time =
-      std::chrono::duration_cast<std::chrono::milliseconds>(
-          std::chrono::seconds(retry_settings.timeout));
-  for (int i = 1; i <= retry_settings.max_attempts && !context.IsCancelled() &&
-                  accumulated_wait_time < max_wait_time;
-       i++) {
-    if (StatusSuccess(response.status)) {
-      return response;
-    }
+  CancellationToken cancellation_token =
+      merge ? CancellationToken([=] { pending_requests->Cancel(url, call_id); })
+            : CancellationToken([=] { request_ptr->Cancel(call_id); });
 
-    if (!retry_settings.retry_condition(response)) {
-      return response;
-    }
+  if (!merge || !IsPending(request_ptr)) {
+    const auto& retry_settings = settings_.retry_settings;
+    auto network = settings_.network_request_handler;
+    auto request_settings = GetRequestSettings(retry_settings);
 
-    // do the periodical sleep and check for cancellation status in between.
-    auto duration_to_sleep =
-        std::min(backdown_period, max_wait_time - accumulated_wait_time);
-    accumulated_wait_time += duration_to_sleep;
+    ExecuteSingleRequest(
```

Comment:
Is this enough? What about the retry and the backdown strategy? I don't see ExecuteSingleRequest doing it.